File: general_algorithm/backtracking/0047_permutations_II.js
/*
Given a collection of numbers, nums, that might contain duplicates, 
return all possible unique permutations in any order.
*/
var permuteUnique = function (nums) {
  let result = [];
  let path = [];
  // used[i] represents whether nums[i] is used
  let used = new Array(nums.length).fill(false);
  // sort nums in ascending order
  nums.sort((a, b) => a - b);
  // define backtracking function
  var backtracking = function (nums) {
    if (path.length === nums.length) {
      result.push([...path]);
      return;
    }
    // loop through nums
    for (let i = 0; i < nums.length; i++) {
      // if nums[i] is used, skip
      if (used[i]) continue;
      // for consecutive duplicates, only use the first one
      if (i > 0 && nums[i] === nums[i - 1] && !used[i - 1]) continue;
      path.push(nums[i]);
      used[i] = true;
      backtracking(nums);
      path.pop();
      used[i] = false;
    }
  };
  backtracking(nums);
  return result;
};

console.log(permuteUnique([1, 2, 3])); // [[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]
console.log(permuteUnique([1, 1, 2])); // [[1,1,2],[1,2,1],[2,1,1]]

File: general_algorithm/backtracking/0047_permutations_II.py
# Given a collection of numbers, nums, that might contain duplicates, return all possible unique permutations in any order.

# https://leetcode.com/problems/permutations-ii/description/
# same code could be used to solve LC46 by deleting line 21 ~ 23

class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        res, path, l = [], [], len(nums)
        used = [False] * l
        nums.sort()
        def backtracking(nums):
            # base case
            if len(path) == l:
                res.append(path[:])
                return
            # recursive case
            for i in range(l):
                # if this element has been used
                if used[i]:
                    continue
                # If nums[i - 1] hasn’t been used, it means that starting a new permutation branch with nums[i] would result in the same permutation as starting with nums[i - 1]. Therefore, we skip it to avoid duplicates.
                if i > 0 and nums[i] == nums[i - 1] and not used[i - 1]:
                    continue
                path.append(nums[i])
                used[i] = True
                backtracking(nums)
                path.pop()
                used[i] = False
                
        backtracking(nums)
        return res